TaskLayer/SearchTask/SearchTask.cs

```diff
@@ -453,5 +454,81 @@ private static MassDiffAcceptor ParseSearchMode(string text)
 
             return massDiffAcceptor;
         }
+
+        /// <summary>
+        /// Handle ambiguity when two theoretical proteins in the 
+        /// search space have the same accession number.
+        /// The accession must be unique for indexed searches
+        /// </summary>
+        /// <param name="proteins"></param>
+        /// <param name="tcAmbiguity"></param>
+        protected static void SanitizeProteinDatabase(List<Protein> proteins, TargetContaminantAmbiguity tcAmbiguity)
+        {
```

Comment:
tbh this section of code seems complicated and error prone... i wrote a simpler version that i think is easier to follow, see what you think:

```
protected static void SanitizeProteinDatabase(List<Protein> proteins, TargetContaminantAmbiguity tcAmbiguity)
        {
            foreach (var accessionGroup in proteins.GroupBy(p => p.Accession))
            {
                int numProteinsWithSameAccession = accessionGroup.Count();

                if (numProteinsWithSameAccession == 1)
                {
                    continue;
                }

                List<Protein> proteinsWithThisAccession = accessionGroup.ToList();
                List<Protein> proteinsToRemove = new List<Protein>();

                if (tcAmbiguity == TargetContaminantAmbiguity.RemoveContaminant)
                {
                    // remove contaminants
                    proteinsToRemove = proteinsWithThisAccession.Where(p => p.IsContaminant).ToList();
                }
                else if (tcAmbiguity == TargetContaminantAmbiguity.RemoveTarget)
                {
                    // remove targets
                    proteinsToRemove = proteinsWithThisAccession.Where(p => !p.IsDecoy && !p.IsContaminant).ToList();
                }

                foreach (var protein in proteinsToRemove)
                {
                    proteins.RemoveAll(p => p == protein);
                    proteinsWithThisAccession.RemoveAll(p => p == protein);
                }

                proteinsToRemove.Clear();

                // if there are still duplicate proteins, rename them to avoid duplicate accessions
                if (proteinsWithThisAccession.Count > 1)
                {
                    int proteinNumber = 1;

                    foreach (var originalProtein in proteinsWithThisAccession)
                    {
                        var renamedProtein = new Protein(originalProtein.BaseSequence, originalProtein + "_" + proteinNumber.ToString(), originalProtein.Organism,
                            originalProtein.GeneNames.ToList(), originalProtein.OneBasedPossibleLocalizedModifications, originalProtein.ProteolysisProducts.ToList(), originalProtein.Name, originalProtein.FullName,
                            originalProtein.IsDecoy, originalProtein.IsContaminant, originalProtein.DatabaseReferences.ToList(), originalProtein.SequenceVariations.ToList(), originalProtein.AppliedSequenceVariations,
                            originalProtein.SampleNameForVariants, originalProtein.DisulfideBonds.ToList(), originalProtein.SpliceSites.ToList(), originalProtein.DatabaseFilePath);

                        proteins.Add(renamedProtein);
                        proteinsToRemove.Add(originalProtein);
                        proteinNumber++;
                    }
                }

                foreach (var protein in proteinsToRemove)
                {
                    proteins.RemoveAll(p => p == protein);
                    proteinsWithThisAccession.RemoveAll(p => p == protein);
                }
            }
        }
```

Reply:
also does this belong in the search task? shouldn't this live in MetaMorpheusTask and then all the tasks would call this method upon loading in the protein DB?

Reply:
That's a point I had considered and am open to doing. I ended up putting it in the search task because I didn't think it substantially influenced the other tasks. I could see a case for XLSearchTask, though. I can put it to the MetaMorpheusTask if you would prefer.

Reply:
I think that's a good idea - as far as I know this mostly affects indexed searches (modern search, SNES, XL search, and probably glyco search). since those compose 3 different tasks it's probably a good idea for them all to reference this block of code

Reply:
I like the revision. Is there a reason to use RemoveAll vs Remove? Shouldn't there only be one instance?

Reply:
Ahh, it keeps decoys for both, though, which I don't like. Hm...

Reply:
yes that is a rather strange quirk. the `RemoveAll(p => p == protein)` forces the use of reference equality and not value equality (`==` and not `.Equals()`). if you use `.Remove()`, then the equals method for the protein is used, which will remove the first protein in the list with the given accession and base sequence - the IsContaminant bool is not used to assess equality. We could fix that in mzLib but this is just a quick-and-dirty fix